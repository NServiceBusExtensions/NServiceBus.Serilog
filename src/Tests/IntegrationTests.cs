using Serilog.Exceptions;
using TypeNameConverter = NServiceBus.Serilog.TypeNameConverter;

[UsesVerify]
public class IntegrationTests
{
    static List<LogEvent> logs;

    static IntegrationTests()
    {
        logs = new();
        var eventSink = new EventSink(logs.Add);

        var configuration = new LoggerConfiguration();
        var enrich = configuration.Enrich;
        enrich.WithExceptionDetails();
        enrich.WithNsbExceptionDetails();
        configuration.MinimumLevel.Verbose();
        configuration.WriteTo.Sink(eventSink);
        Log.Logger = configuration.CreateLogger();
        LogManager.Use<SerilogFactory>();
    }

//#if NETCOREAPP3_1
//    [Fact]
//    public async Task WriteStartupDiagnostics()
//    {
//        var events = await Send(
//            new StartHandler
//            {
//                Property = "TheProperty"
//            });
//        await Verify<StartupDiagnostics>(events);
//    }
//#endif

    [Fact]
    public async Task Handler()
    {
        var events = await Send(
            new StartHandler
            {
                Property = "TheProperty"
            });
        await Verify<StartHandler>(events);
    }

    [Fact]
    public async Task GenericHandler()
    {
        var events = await Send(
            new StartGenericHandler<string>
            {
                Property = "TheProperty"
            });
        await Verify<StartGenericHandler<string>>(events);
    }

    [Fact]
    public async Task WithCustomHeader()
    {
        var events = await Send(
            new StartHandler
            {
                Property = "TheProperty"
            },
            options => options.SetHeader("CustomHeader", "CustomValue"));
        await Verify<StartHandler>(events);
    }

    [Fact]
    public async Task WithConvertedCustomHeader()
    {
        var events = await Send(
            new StartHandler
            {
                Property = "TheProperty"
            }, options => options.SetHeader("ConvertHeader", "CustomValue"));
        await Verify<StartHandler>(events);
    }

    //[Fact]
    //public async Task SagaNotFound()
    //{
    //    var events = await Send(
    //        new NotFoundSagaMessage(),
    //        options =>
    //        {
    //            options.SetHeader(Headers.SagaId, Guid.NewGuid().ToString());
    //            options.SetHeader(Headers.SagaType, typeof(NotFoundSaga).FullName);
    //        });
    //    await Verify<NotFoundSagaMessage>(events);
    //}

    [Fact]
    public async Task HandlerThatLogs()
    {
        var events = await Send(new StartHandlerThatLogs());
        await Verify<StartHandlerThatLogs>(events);
    }

    [Fact]
    public async Task HandlerThatThrows()
    {
        var events = await Send(
            new StartHandlerThatThrows
            {
                Property = "TheProperty"
            });
        await Verify<StartHandlerThatThrows>(events)
            .ScrubLinesContaining(
                "Handler start time",
                "Handler failure time");
    }

    [Fact]
    public async Task Saga()
    {
        var events = await Send(
            new StartSaga
            {
                Property = "TheProperty"
            });
        var logEvents = events.ToList();
        await Verify<StartSaga>(logEvents);
    }

    [Fact]
    public async Task BehaviorThatThrows()
    {
        var events = await Send(
            new StartBehaviorThatThrows
            {
                Property = "TheProperty"
            },
            extraConfiguration: _ => _.EnableFeature<BehaviorThatThrowsFeature>());
        var logEvents = events.ToList();
        await Verify<StartBehaviorThatThrows>(logEvents);
    }

    static SettingsTask Verify<T>(IEnumerable<LogEventEx> logEvents)
    {
        var list = logEvents.ToList();
        var logsForTarget = list.LogsForType<T>().ToList();
        return Verifier.Verify(
            new
            {
                logsForTarget,
                logsForNsbSerilog = list.LogsForNsbSerilog().ToList(),
                logsWithExceptions = list.LogsWithExceptions().ToList()
            });
    }

    static async Task<IEnumerable<LogEventEx>> Send(
        object message,
        Action<SendOptions>? optionsAction = null,
        Action<EndpointConfiguration>? extraConfiguration = null)
    {
        logs.Clear();
        var suffix = TypeNameConverter.GetName(message.GetType())
            .Replace("<","_")
            .Replace(">","_");
        var configuration = ConfigBuilder.BuildDefaultConfig("SerilogTests" + suffix);
        configuration.PurgeOnStartup(true);
        extraConfiguration?.Invoke(configuration);

        var serilogTracing = configuration.EnableSerilogTracing();
        serilogTracing.EnableSagaTracing();
        serilogTracing.UseHeaderConversion((key, _) =>
        {
            if (key == "ConvertHeader")
            {
                return new("NewKey", new ScalarValue("newValue"));
            }

            return null;
        });
        serilogTracing.EnableMessageTracing();
        var resetEvent = new ManualResetEvent(false);
        configuration.RegisterComponents(components => components.RegisterSingleton(resetEvent));

        var recoverability = configuration.Recoverability();
        recoverability.Delayed(settings =>
        {
            settings.TimeIncrease(TimeSpan.FromMilliseconds(1));
            settings.NumberOfRetries(1);
        });
        recoverability.Immediate(settings => { settings.NumberOfRetries(1); });

        recoverability.Failed(settings => settings
            .OnMessageSentToErrorQueue(_ =>
            {
                resetEvent.Set();
                return Task.CompletedTask;
            }));

        var endpoint = await Endpoint.Start(configuration);
        var sendOptions = new SendOptions();
        optionsAction?.Invoke(sendOptions);
        sendOptions.SetMessageId("00000000-0000-0000-0000-000000000001");
        sendOptions.RouteToThisEndpoint();
        await endpoint.Send(message, sendOptions);
        if (!resetEvent.WaitOne(TimeSpan.FromSeconds(10)))
        {
            throw new("No Set received.");
        }

        await endpoint.Stop();

        return logs.Select(x =>
            new LogEventEx
            (
                messageTemplate: x.MessageTemplate,
                level: x.Level,
                properties: x.Properties,
                exception: x.Exception
            ));
    }
}