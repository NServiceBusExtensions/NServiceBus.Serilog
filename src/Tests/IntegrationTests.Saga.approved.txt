{
  logsForTarget: [
    {
      MessageTemplate: {
        Text: 'Hello from {@Saga}. Message: {@Message}'
      },
      Level: 'Information',
      Properties: {
        Saga: {
          Value: 'TheSaga'
        },
        Message: {
          TypeTag: 'StartSaga',
          Properties: [
            {
              Name: 'Property',
              Value: {
                Value: 'TheProperty'
              }
            }
          ]
        },
        Handler: {
          Value: 'TheSaga'
        },
        IncomingMessageId: {
          Value: 'Guid 1'
        },
        IncomingMessageType: {
          Value: 'StartSaga'
        },
        CorrelationId: {
          Value: 'Guid 1'
        },
        ConversationId: {
          Value: 'Guid 2'
        },
        SourceContext: {
          Value: 'StartSaga'
        },
        ProcessingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        }
      }
    },
    {
      MessageTemplate: {
        Text: 'Receive message {IncomingMessageType} {IncomingMessageId}.'
      },
      Level: 'Information',
      Properties: {
        IncomingMessage: {
          TypeTag: 'StartSaga',
          Properties: [
            {
              Name: 'Property',
              Value: {
                Value: 'TheProperty'
              }
            }
          ]
        },
        OriginatingHostId: {
          Value: 'Guid 3'
        },
        MessageIntent: {
          Value: 'Send'
        },
        OriginatingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        },
        OriginatingMachine: {
          Value: 'MachineName'
        },
        ReplyToAddress: {
          Value: 'SerilogTestsStartSaga'
        },
        IncomingMessageId: {
          Value: 'Guid 1'
        },
        IncomingMessageType: {
          Value: 'StartSaga'
        },
        CorrelationId: {
          Value: 'Guid 1'
        },
        ConversationId: {
          Value: 'Guid 2'
        },
        SourceContext: {
          Value: 'StartSaga'
        },
        ProcessingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        }
      }
    },
    {
      MessageTemplate: {
        Text: 'Saga execution \'{SagaType}\' \'{SagaId}\'.'
      },
      Level: 'Information',
      Properties: {
        SagaType: {
          Value: 'TheSaga'
        },
        SagaId: {
          Value: 'Guid 4'
        },
        StartTime: {
          Value: 'DateTimeOffset 1'
        },
        FinishTime: {
          Value: 'DateTimeOffset 2'
        },
        IsCompleted: {
          Value: false
        },
        IsNew: {
          Value: true
        },
        Initiator: {
          TypeTag: 'SagaChangeInitiator',
          Properties: [
            {
              Name: 'InitiatingMessageId',
              Value: {
                Value: 'Guid 1'
              }
            },
            {
              Name: 'MessageType',
              Value: {
                Value: 'StartSaga'
              }
            },
            {
              Name: 'IsSagaTimeoutMessage',
              Value: {
                Value: false
              }
            },
            {
              Name: 'TimeSent',
              Value: {
                Value: 'DateTime 1'
              }
            },
            {
              Name: 'OriginatingMachine',
              Value: {
                Value: 'MachineName'
              }
            },
            {
              Name: 'OriginatingEndpoint',
              Value: {
                Value: 'SerilogTestsStartSaga'
              }
            },
            {
              Name: 'Intent',
              Value: {
                Value: 'Send'
              }
            }
          ]
        },
        ResultingMessages: {
          Elements: [
            {
              TypeTag: 'SagaChangeOutput',
              Properties: [
                {
                  Name: 'MessageType',
                  Value: {
                    Value: 'BackIntoSaga'
                  }
                },
                {
                  Name: 'Destination',
                  Value: {
                    Value: 'SerilogTestsStartSaga'
                  }
                },
                {
                  Name: 'ResultingMessageId',
                  Value: {
                    Value: 'Guid 5'
                  }
                },
                {
                  Name: 'MessageIntent',
                  Value: {
                    Value: 'Send'
                  }
                }
              ]
            }
          ]
        },
        Entity: {
          TypeTag: 'TheSagaData',
          Properties: [
            {
              Name: 'Property',
              Value: {
                Value: 'TheProperty'
              }
            },
            {
              Name: 'Id',
              Value: {
                Value: 'Guid 4'
              }
            },
            {
              Name: 'Originator',
              Value: {
                Value: 'SerilogTestsStartSaga'
              }
            },
            {
              Name: 'OriginalMessageId',
              Value: {
                Value: 'Guid 1'
              }
            }
          ]
        },
        IncomingMessageId: {
          Value: 'Guid 1'
        },
        IncomingMessageType: {
          Value: 'StartSaga'
        },
        CorrelationId: {
          Value: 'Guid 1'
        },
        ConversationId: {
          Value: 'Guid 2'
        },
        SourceContext: {
          Value: 'StartSaga'
        },
        ProcessingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        }
      }
    },
    {
      MessageTemplate: {
        Text: 'Sent message {OutgoingMessageType} {OutgoingMessageId}.'
      },
      Level: 'Information',
      Properties: {
        OutgoingMessage: {
          TypeTag: 'StartSaga',
          Properties: [
            {
              Name: 'Property',
              Value: {
                Value: 'TheProperty'
              }
            }
          ]
        },
        UnicastRoutes: {
          Elements: [
            {
              Value: 'SerilogTestsStartSaga'
            }
          ]
        },
        OriginatingHostId: {
          Value: 'Guid 3'
        },
        MessageIntent: {
          Value: 'Send'
        },
        OriginatingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        },
        OriginatingMachine: {
          Value: 'MachineName'
        },
        ReplyToAddress: {
          Value: 'SerilogTestsStartSaga'
        },
        OutgoingMessageId: {
          Value: 'Guid 1'
        },
        OutgoingMessageType: {
          Value: 'StartSaga'
        },
        CorrelationId: {
          Value: 'Guid 1'
        },
        ConversationId: {
          Value: 'Guid 2'
        },
        SourceContext: {
          Value: 'StartSaga'
        },
        ProcessingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        }
      }
    },
    {
      MessageTemplate: {
        Text: 'Sent message {OutgoingMessageType} {OutgoingMessageId}.'
      },
      Level: 'Information',
      Properties: {
        OutgoingMessage: {
          TypeTag: 'BackIntoSaga',
          Properties: [
            {
              Name: 'Property',
              Value: {
                Value: 'TheProperty'
              }
            }
          ]
        },
        UnicastRoutes: {
          Elements: [
            {
              Value: 'SerilogTestsStartSaga'
            }
          ]
        },
        OriginatingHostId: {
          Value: 'Guid 3'
        },
        MessageIntent: {
          Value: 'Send'
        },
        OriginatingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        },
        OriginatingMachine: {
          Value: 'MachineName'
        },
        OriginatingSagaId: {
          Value: 'Guid 4'
        },
        OriginatingSagaType: {
          Value: 'TheSaga, Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=ce8ec7717ba6fbb6'
        },
        RelatedTo: {
          Value: 'Guid 1'
        },
        ReplyToAddress: {
          Value: 'SerilogTestsStartSaga'
        },
        Handler: {
          Value: 'TheSaga'
        },
        IncomingMessageId: {
          Value: 'Guid 1'
        },
        IncomingMessageType: {
          Value: 'StartSaga'
        },
        CorrelationId: {
          Value: 'Guid 1'
        },
        ConversationId: {
          Value: 'Guid 2'
        },
        SourceContext: {
          Value: 'StartSaga'
        },
        ProcessingEndpoint: {
          Value: 'SerilogTestsStartSaga'
        }
      }
    }
  ]
}
